lab/src/Microsoft.ServiceModel.Syndication/src/XmlReaderWrapper.cs

```diff
@@ -0,0 +1,575 @@
+                        if (reader.IsEmptyElement)
+                        {
+                            writer.WriteEndElement();
+                            break;
+                        }
+                        break;
+                    case XmlNodeType.Text:
+                        if (canReadChunk)
+                        {
+                            if (writeNodeBuffer == null)
+                            {
+                                writeNodeBuffer = new char[WriteNodeBufferSize];
+                            }
+                            int read;
+                            while ((read = reader.ReadValueChunk(writeNodeBuffer, 0, WriteNodeBufferSize)) > 0)
+                            {
+                                writer.WriteChars(writeNodeBuffer, 0, read);
+                            }
+                        }
+                        else
+                        {
+                            writer.WriteString(await reader.GetValueAsync());
+                        }
+                        break;
+
+                    case XmlNodeType.Whitespace:
+                    case XmlNodeType.SignificantWhitespace:
+                        writer.WriteWhitespace(await reader.GetValueAsync());
+                        break;
+
+                    case XmlNodeType.CDATA:
+                        writer.WriteCData(await reader.GetValueAsync());
+                        break;
+
+                    case XmlNodeType.EntityReference:
+                        writer.WriteEntityRef(reader.Name);
+                        break;
+
+                    case XmlNodeType.XmlDeclaration:
+                    case XmlNodeType.ProcessingInstruction:
+                        writer.WriteProcessingInstruction(reader.Name, await reader.GetValueAsync());
+                        break;
+
+                    case XmlNodeType.DocumentType:
+                        writer.WriteDocType(reader.Name, reader.GetAttribute("PUBLIC"), reader.GetAttribute("SYSTEM"), await reader.GetValueAsync());
+                        break;
+
+                    case XmlNodeType.Comment:
+                        writer.WriteComment(await reader.GetValueAsync());
+                        break;
+
+                    case XmlNodeType.EndElement:
+                        writer.WriteFullEndElement();
+                        break;
+                }
+            } while (await reader.ReadAsync() && (d < reader.Depth || (d == reader.Depth && reader.NodeType == XmlNodeType.EndElement)));
+        }
+    }
+
+    internal static class XmlReaderExtensions
+    {
+        public static async Task ReadStartElementAsync(this XmlReader reader)
+        {
+            if (await reader.MoveToContentAsync() != XmlNodeType.Element)
+            {
+                throw new InvalidOperationException(reader.NodeType.ToString() + " is an invalid XmlNodeType");
+            }
+
+            reader.Read();
```

Comment:
The original code was reader.ReadAsync(). I changed it to read() just to keep the ReaderExtension working but I think this approach will give us some problems. If the reader has no support for async and arrives to this point it will crash.

The reader will not always have async support, it depends on the user.